### PR TITLE
A pile of Theming nits fixes

### LINKED ISSUE #13456: nits from themes PR
see #12992 

________________________________________
* [ ] In `src/cascadia/TerminalSettingsModel/Theme.idl`:
```diff
> +        Color,
+        TerminalBackground
+    };
+
+    runtimeclass ThemeColor
+    {
+        ThemeColor();
+        static ThemeColor FromColor(Microsoft.Terminal.Core.Color color);
+        static ThemeColor FromAccent();
+        static ThemeColor FromTerminalBackground();
+
+        Microsoft.Terminal.Core.Color Color { get; };
+        ThemeColorType ColorType;
+
+        static Microsoft.Terminal.Core.Color ColorFromBrush(Windows.UI.Xaml.Media.Brush brush);
+        Windows.UI.Xaml.Media.Brush Evaluate(Windows.UI.Xaml.ResourceDictionary res,
```
XAML in the settings model -- this will cause trouble for a consumer like the shell extension in the future.

* [ ] In `src/cascadia/TerminalSettingsEditor/GlobalAppearance.cpp`:
```diff
> +        // Surprisingly, though this is called every time we navigate to the page,
+        // the list does not keep growing on each navigation.
```
This comment has a VERY threatening aura! I am almost certain it will break at some point...
_____________
* [x] In `src/cascadia/TerminalApp/AppLogic.cpp`:
```diff
> @@ -964,7 +960,7 @@ namespace winrt::TerminalApp::implementation
 
     void AppLogic::_RefreshThemeRoutine()
     {
-        _ApplyTheme(_settings.GlobalSettings().Theme());
+        _ApplyTheme(GetRequestedTheme());
```
I feel like ApplyTheme should take a Theme ;P
________________________________________
* [x] In `src/cascadia/TerminalApp/TerminalPage.cpp`:
```diff
> +        //
+        // This helper allows us to instead lookup the value of a resource
+        // specified by `key` for the given `requestedTheme`, from the
+        // dictionaries in App.xaml. Make sure the value is actually there!
+        // Otherwise this'll throw like any other Lookup for a resource that
+        // isn't there.
+        static const auto lookup = [](auto& res, auto& requestedTheme, auto& key) {
+            // You want the Default version of the resource? Great, the App is
+            // always in the OS theme. Just look it up and be done.
+            if (requestedTheme == ElementTheme::Default)
+            {
+                return res.Lookup(key);
+            }
+            static const auto lightKey = winrt::box_value(L"Light");
+            static const auto darkKey = winrt::box_value(L"Dark");
+            // There isn't an ElementTheme::HighContrast.
```
so... what exactly happens in HC then?
________________________________________
* [ ] In `src/cascadia/TerminalSettingsModel/Theme.cpp`:
```diff
> +        {                                                                                      \
+            return json.isObject();                                                            \
+        }                                                                                      \
+                                                                                               \
+        Json::Value ToJson(const nameSpace::name& val)                                         \
+        {                                                                                      \
+            if (val == nullptr)                                                                \
+                return Json::Value::null;                                                      \
+            Json::Value json{ Json::ValueType::objectValue };                                  \
+            macro(THEME_SETTINGS_TO_JSON);                                                     \
+            return json;                                                                       \
+        }                                                                                      \
+                                                                                               \
+        std::string TypeDescription() const                                                    \
+        {                                                                                      \
+            return "name (You should never see this)";                                         \
```
lol this will actually say the word "name" since it's not in a macro-stringify expression

<details>
<summary>resolved</summary>
________________________________________
* [x] Does this work with background images?

Yep, sure does.

________________________________________
* [x] It looks insane to have a light terminal BG titlebar, and in dark OS theme - the caption btns are also white:

image

________________________________________
* [x] In `src/cascadia/LocalTests_SettingsModel/ThemeTests.cpp`:
```diff
> +                "background": "#FFFF8800",
+                "unfocusedBackground": "#FF884400"
```
wait but utils parses RGBA
how is this test passing?

wait, i am so confused
___________________________
* [x] In `src/cascadia/TerminalSettingsModel/Theme.h`:
```diff
> +
+        static com_ptr<Theme> FromJson(const Json::Value& json);
+        void LayerJson(const Json::Value& json);
+        Json::Value ToJson() const;
+
+        winrt::Windows::UI::Xaml::ElementTheme RequestedTheme() const noexcept;
+
+        WINRT_PROPERTY(winrt::hstring, Name);
+
+        MTSM_THEME_SETTINGS(THEME_SETTINGS_INITIALIZE)
+
+    private:
+    };
+
+#undef THEME_SETTINGS_INITIALIZE
+#undef THEME_SETTINGS_COPY
```
do we need to undef COPY_THEME...?
________________________________________
* [x] In `src/cascadia/TerminalSettingsModel/TerminalSettingsSerializationHelpers.h`:
```diff
> +        }
+        case winrt::Microsoft::Terminal::Settings::Model::ThemeColorType::Color:
+        {
+            return til::u16u8(til::color{ val.Color() }.ToHexString(false));
+        }
+        case winrt::Microsoft::Terminal::Settings::Model::ThemeColorType::TerminalBackground:
+        {
+            return "terminalBackground";
+        }
+        }
+        return til::u16u8(til::color{ val.Color() }.ToHexString(false));
+    }
+
+    std::string TypeDescription() const
+    {
+        return "ThemeColor (#rrggbb, #rgb, #aarrggbb, accent, terminalBackground)";
```
isn't it rrggbbaa?
________________________________________
* [x] In `src/cascadia/TerminalSettingsModel/TerminalSettingsSerializationHelpers.h`:
```diff
> @@ -547,6 +547,79 @@ JSON_ENUM_MAPPER(::winrt::Microsoft::Terminal::Settings::Model::InfoBarMessage)
     };
 };
 
+template<>
+struct ::Microsoft::Terminal::Settings::Model::JsonUtils::ConversionTrait<winrt::Microsoft::Terminal::Settings::Model::ThemeColor>
+{
+    winrt::Microsoft::Terminal::Settings::Model::ThemeColor FromJson(const Json::Value& json)
+    {
+        if (json == Json::Value::null)
+        {
+            return nullptr;
+        }
+        const auto string{ Detail::GetStringView(json) };
+        if (string == "accent")
```
nit: prefer declaring these magic constant values in the class body and referencing them later
________________________________________
* [x] In `src/cascadia/TerminalSettingsModel/CascadiaSettingsSerialization.cpp`:
```diff
> @@ -1092,6 +1113,20 @@ Json::Value CascadiaSettings::ToJson() const
     }
     json[JsonKey(SchemesKey)] = schemes;
 
+    Json::Value themes{ Json::ValueType::arrayValue };
+    for (const auto& entry : _globals->Themes())
+    {
+        // Ignore the built in themes, when serializing the themes back out. We
```
implicit contract: users cannot overwrite system themes
________________________________________
* [x] In `src/cascadia/TerminalApp/AppLogic.h`:
```diff
>          // -------------------------------- WinRT Events ---------------------------------
+        // PropertyChanged is surprisingly not a typed event, so we'll define that one manually.
```
W A T

do we do that everywhere? if not, why not? Should we have done here what we do elsewhere?

________________________________________
* [x] In `src/cascadia/TerminalSettingsEditor/GlobalAppearance.idl`:
```diff
> @@ -21,7 +21,8 @@ namespace Microsoft.Terminal.Settings.Editor
         IInspectable CurrentLanguage;
 
         IInspectable CurrentTheme;
-        Windows.Foundation.Collections.IObservableVector<Microsoft.Terminal.Settings.Editor.EnumEntry> ThemeList { get; };
+        static String ThemeNameConverter(Microsoft.Terminal.Settings.Model.Theme theme);
+        Windows.Foundation.Collections.IObservableVector<Microsoft.Terminal.Settings.Model.Theme> ThemeList { get; };
```
@PankajBhojwani FYI since this will impact MVVM -- this is a raw exposure of a settings model object
________________________________________
* [x] In `src/cascadia/TerminalSettingsEditor/GlobalAppearance.cpp`:
```diff
> +    constexpr std::wstring_view systemThemeName{ L"system" };
+    constexpr std::wstring_view darkThemeName{ L"dark" };
+    constexpr std::wstring_view lightThemeName{ L"light" };
```
these became global constants in TSE, but not in TSM -- curious!
________________________________________
* [x] In `src/cascadia/TerminalApp/TerminalPage.cpp`:
```diff
> +        _activated = activated;
+        _updateTabRowColors();
+    }
```
so _activated is ignored?
________________________________________
* [x] In `src/cascadia/TerminalSettingsModel/Theme.cpp`:
```diff
> +    HKEY hKey{ nullptr };
+    if (RegOpenKeyEx(HKEY_CURRENT_USER, RegKeyDwm, 0, KEY_READ, &hKey) == ERROR_SUCCESS)
+    {
+        return wil::unique_hkey{ hKey };
+    }
+    return nullptr;
+}
+static DWORD readDwmSubValue(const wil::unique_hkey& dwmRootKey, const wchar_t* key)
+{
+    DWORD val{ 0 };
+    DWORD size{ sizeof(val) };
+    LOG_IF_FAILED(RegQueryValueExW(dwmRootKey.get(), key, nullptr, nullptr, reinterpret_cast<BYTE*>(&val), &size));
+    return val;
+}
+
+static til::color _getAccentColorForTitlebar()
```
do we reload this when the accent color changes?
________________________________________
________________________________________
* [x] In `src/cascadia/TerminalSettingsModel/Theme.cpp`:
```diff
> +    }
+
+    return theme;
+}
+
+// Method Description:
+// - Create a new instance of this class from a serialized JsonObject.
+// Arguments:
+// - json: an object which should be a serialization of a ColorScheme object.
+// Return Value:
+// - Returns nullptr for invalid JSON.
+winrt::com_ptr<Theme> Theme::FromJson(const Json::Value& json)
+{
+    auto result = winrt::make_self<Theme>();
+
+    if (json.isString())
```
we control the entire themes array, from its inception to its death. Who would put a string in it?
________________________________________
* [x] In `src/cascadia/TerminalApp/TerminalPage.cpp`:
```diff
> +                    if (winrt::unbox_value<winrt::hstring>(dictionaryKey) !=
+                        winrt::unbox_value<winrt::hstring>(requestedThemeKey))
```
but do we need users to crash if this somehow happens in production?
________________________________________
* [x] In `src/cascadia/TerminalApp/TerminalPage.cpp`:
```diff
> +                }
+            }
+
+            // We didn't find it in the requested dict, fall back to the default dictionary.
+            return res.Lookup(key);
+        };
+
+        // Use our helper to lookup the theme-aware version of the resource.
+        const auto tabViewBackgroundKey = winrt::box_value(L"TabViewBackground");
+        const auto backgroundSolidBrush = lookup(res, requestedTheme, tabViewBackgroundKey).as<Media::SolidColorBrush>();
+
+        til::color bgColor = backgroundSolidBrush.Color();
+
+        if (_settings.GlobalSettings().UseAcrylicInTabRow())
+        {
+            const til::color backgroundColor = backgroundSolidBrush.Color();
```
You marked it as resolved... but the line of code is still here.

</details>

<details>
<summary>follow ups</summary>
________________________________________
* [ ] In `src/cascadia/TerminalSettingsModel/CascadiaSettingsSerialization.cpp`:
```diff
> @@ -531,6 +532,25 @@ void SettingsLoader::_parse(const OriginTag origin, const winrt::hstring& source
         }
     }
 
+    {
+        for (const auto& themeJson : json.themes)
+        {
+            if (const auto theme = Theme::FromJson(themeJson))
+            {
+                if (origin != OriginTag::InBox &&
```
follow up: we should add an origin tag to themes, after #12800 lands.
________________________________________
* [ ] In `src/cascadia/TerminalSettingsEditor/GlobalAppearance.xaml`:
```diff
>                            ItemsSource="{x:Bind ThemeList, Mode=OneWay}"
                           SelectedItem="{x:Bind CurrentTheme, Mode=TwoWay}"
-                          Style="{StaticResource ComboBoxSettingStyle}" />
+                          Style="{StaticResource ComboBoxSettingStyle}">
+                    <ComboBox.ItemTemplate>
+                        <DataTemplate x:DataType="model:Theme">
+                            <TextBlock Text="{x:Bind local:GlobalAppearance.ThemeNameConverter((model:Theme)), Mode=OneWay}" />
```
@carlos-zamora, can you make sure this is still accessible? we have had issues with putting custom data templates into combo boxes.
________________________________________

* [ ] In `src/types/utils.cpp`:
```diff
>      }
-    else
+    else if (str.size() == 7)
+    {
+        rStr = std::string(&str.at(1), 2);
+        gStr = std::string(&str.at(3), 2);
+        bStr = std::string(&str.at(5), 2);
+        aStr = "ff";
```
I don't love that we force a string parser to parse ff just to get 255... ;P
________________________________________
</details>

### PR DESCRIPTION
**Pull request by zadjii-msft**
The main fix here is for the caption button colors. If you had a dark OS/app theme, and a light titlebar, we'd end up with light glyphs, so the caption buttons would be impossible to find.

There's also a pile of nits from #12992 in here. Probably enough to close #13456 out, but I'll let Dustin be the judge. 

Filing today, to get in for 1.16 selfhost (@DHowett)

### CODE REVIEW

File: src/cascadia/TerminalApp/TitlebarControl.cpp [hunk 1]
```diff
@@ -144,4 +146,20 @@ namespace winrt::TerminalApp::implementation
...
+    {
+        til::color c;
+        if (auto acrylic = brush.try_as<winrt::Windows::UI::Xaml::Media::AcrylicBrush>())
+        {
+            c = acrylic.TintColor();
+        }
+        else if (auto solidColor = brush.try_as<winrt::Windows::UI::Xaml::Media::SolidColorBrush>())
+        {
+            c = solidColor.Color();
+        }
```
Comment (DHowett) [inline src/cascadia/TerminalApp/TitlebarControl.cpp:159]: what will `c` be initialized to if it's neither Acrylic nor SolidColor?

File: src/cascadia/WindowsTerminal/NonClientIslandWindow.cpp [hunk 2]
```diff
@@ -1133,4 +1133,5 @@ bool NonClientIslandWindow::_IsTitlebarVisible() const
 void NonClientIslandWindow::SetTitlebarBackground(winrt::Windows::UI::Xaml::Media::Brush brush)
 {
     _titlebar.Background(brush);
+    _titlebar.SetBackgroundBrush(brush);
```
Comment (DHowett) [inline src/cascadia/WindowsTerminal/NonClientIslandWindow.cpp:1136]: Should `.Background()` just be inside `SetBackgroundBrush`? Also, there's now _two setters for the background brush :/_

Comment (zadjii-msft) [reply src/cascadia/WindowsTerminal/NonClientIslandWindow.cpp:1136]: 1. maybe
2. gods it's horrible: cause TitlebarControl is a Grid, it has a Background method, but I can't override that, so 🤷

Comment (DHowett) [reply src/cascadia/WindowsTerminal/NonClientIslandWindow.cpp:1136]: Okay weird question . . . can you add a listener for `PropertyChanged` entirely inside TitleBarControl? It listens to itself for Background changes

Comment (DHowett) [reply src/cascadia/WindowsTerminal/NonClientIslandWindow.cpp:1136]: And then you don't need to add another setter OR call a different setter :O

**Review by lhecker — APPROVED** [review]
(I only have the same question as Dustin regarding `til::color`.)

File: src/cascadia/LocalTests_SettingsModel/ThemeTests.cpp [hunk 3]
```diff
@@ -72,7 +76,8 @@ namespace SettingsModelLocalTests
         VERIFY_IS_NOT_NULL(theme->TabRow());
         VERIFY_IS_NOT_NULL(theme->TabRow().Background());
         VERIFY_ARE_EQUAL(Settings::Model::ThemeColorType::Color, theme->TabRow().Background().ColorType());
-        VERIFY_ARE_EQUAL(rgb(0xff, 0x88, 0x00), theme->TabRow().Background().Color());
```
Comment (DHowett) [inline src/cascadia/LocalTests_SettingsModel/ThemeTests.cpp:75]: I still don't understand how these were passing before -- or, weren't they?

**Review by DHowett — APPROVED** [review]

File: src/cascadia/LocalTests_SettingsModel/ThemeTests.cpp (hunk 3 above)
Comment (zadjii-msft) [reply src/cascadia/LocalTests_SettingsModel/ThemeTests.cpp:75]: > or, weren't they

nah they totally weren't

File: src/cascadia/TerminalApp/TerminalPage.cpp [hunk 4]
```diff
@@ -4101,27 +4100,25 @@ namespace winrt::TerminalApp::implementation
 
             TitlebarBrush(acrylicBrush);
         }
-        else if (theme.TabRow())
+        else if (const auto tabRowBg{ _activated ? theme.TabRow().Background() :
```
Comment (PankajBhojwani) [inline src/cascadia/TerminalApp/TerminalPage.cpp:4103]: This call is causing a crash on startup in the current build of main with the error `Access violation reading location 0x0000000000000000` (just did a git pull, clean and tried to run)

Comment (PankajBhojwani) [reply src/cascadia/TerminalApp/TerminalPage.cpp:4103]: Ah, I think the `theme.TabRow() != nullptr` check should come before we try to access the `Background`?

**Pull request merged**